File: Кросс-платформне програмування/lab_№4/task_Java/src/CarParkingManager.java
import java.io.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class CarParkingManager {
    private static final String FILENAME = "car_parking_records.txt";
    private static final List<CarParkingRecord> records = new ArrayList<>();

    public static void main(String[] args) {
        loadDataFromFile();
        Scanner scanner = new Scanner(System.in);

        while (true) {
            System.out.println("\nOptions:");
            System.out.println("1. Add Parking Record");
            System.out.println("2. View All Parking Records");
            System.out.println("3. Search by License Plate");
            System.out.println("4. Search by Owner's Phone");
            System.out.println("5. Filter by Parking Hour");
            System.out.println("6. Check Payment Status");
            System.out.println("7. Exit");
            System.out.print("Enter your choice: ");
            int choice = scanner.nextInt();
            scanner.nextLine();

            switch (choice) {
                case 1:
                    addParkingRecord(scanner);
                    break;
                case 2:
                    viewAllParkingRecords();
                    break;
                case 3:
                    searchByLicensePlate(scanner);
                    break;
                case 4:
                    searchByOwnerPhone(scanner);
                    break;
                case 5:
                    filterByParkingHour(scanner);
                    break;
                case 6:
                    checkPaymentStatus(scanner);
                    break;
                case 7:
                    saveDataToFile();
                    System.out.println("Exiting...");
                    System.exit(0);
                    break;
                default:
                    System.out.println("Invalid choice. Please try again.");
            }
        }
    }

    private static void addParkingRecord(Scanner scanner) {
        System.out.println("\nEnter Car Details:");
        System.out.print("Brand and Model: ");
        String brandModel = scanner.nextLine();
        System.out.print("License Plate: ");
        String licensePlate = scanner.nextLine();
        System.out.print("Owner's Phone: ");
        String ownerPhone = scanner.nextLine();
        System.out.print("Parking Hour: ");
        int parkingHour = scanner.nextInt();
        System.out.print("Prepaid Hours: ");
        int prepaidHours = scanner.nextInt();
        scanner.nextLine();

        CarParkingRecord record = new CarParkingRecord(brandModel, licensePlate, ownerPhone, parkingHour, prepaidHours);
        records.add(record);
        System.out.println("Parking record added successfully.");
    }

    private static void viewAllParkingRecords() {
        if (records.isEmpty()) {
            System.out.println("No parking records available.");
        } else {
            System.out.println("\nAll Parking Records:");
            for (CarParkingRecord record : records) {
                System.out.println(record);
            }
        }
    }

    private static void searchByLicensePlate(Scanner scanner) {
        System.out.print("\nEnter License Plate to search: ");
        String licensePlate = scanner.nextLine();
        boolean found = false;

        for (CarParkingRecord record : records) {
            if (record != null && record.getLicensePlate().equalsIgnoreCase(licensePlate)) {
                System.out.println("Matching Record: " + record);
                found = true;
                break;
            }
        }

        if (!found) {
            System.out.println("No matching record found.");
        }
    }

    private static void searchByOwnerPhone(Scanner scanner) {
        System.out.print("\nEnter Owner's Phone to search: ");
        String ownerPhone = scanner.nextLine();
        boolean found = false;

        for (CarParkingRecord record : records) {
            if (record != null && record.getOwnerPhone().equals(ownerPhone)) {
                System.out.println("Matching Record: " + record);
                found = true;
                break;
            }
        }

        if (!found) {
            System.out.println("No matching record found.");
        }
    }

    private static void filterByParkingHour(Scanner scanner) {
        System.out.print("\nEnter Parking Hour to filter: ");
        int parkingHour = scanner.nextInt();
        boolean found = false;

        for (CarParkingRecord record : records) {
            if (record != null && record.getParkingHour() == parkingHour) {
                System.out.println("Matching Record: " + record);
                found = true;
            }
        }

        if (!found) {
            System.out.println("No matching record found.");
        }
    }

    private static void checkPaymentStatus(Scanner scanner) {
        System.out.print("\nEnter License Plate to check payment status: ");
        String licensePlate = scanner.nextLine();
        boolean found = false;

        for (CarParkingRecord record : records) {
            if (record != null && record.getLicensePlate().equalsIgnoreCase(licensePlate)) {
                if (record.getPrepaidHours() >= (getCurrentHour() - record.getParkingHour())) {
                    System.out.println("No additional payment required. Car can leave.");
                } else {
                    System.out.println("Additional payment required for " + (getCurrentHour() - record.getParkingHour()) + " hours.");
                }
                found = true;
                break;
            }
        }

        if (!found) {
            System.out.println("No matching record found.");
        }
    }

    private static int getCurrentHour() {
        return 10;
    }

    private static void saveDataToFile() {
        try (PrintWriter writer = new PrintWriter(new FileWriter(FILENAME))) {
            for (CarParkingRecord record : records) {
                if (record != null) {
                    writer.println(record.toText());
                }
            }
            System.out.println("Data saved successfully to file.");
        } catch (IOException e) {
            System.out.println("Failed to save data to file: " + e.getMessage());
        }
    }


    private static void loadDataFromFile() {
        records.clear();
        try (Scanner scanner = new Scanner(new FileReader(FILENAME))) {
            while (scanner.hasNextLine()) {
                String line = scanner.nextLine();
                CarParkingRecord record = CarParkingRecord.fromText(line);
                records.add(record);
            }
            System.out.println("Data loaded successfully from file.");
        } catch (IOException e) {
            System.out.println("No existing data found. Starting with an empty record list.");
        }
    }
}

File: Кросс-платформне програмування/lab_№4/task_Kotlin/src/CarParkingManager.kt
import java.io.File
import java.util.Scanner

class CarParkingManager {
    companion object {
        private const val FILENAME = "car_parking_records.txt"
        private val records = mutableListOf<CarParkingRecord>()

        @JvmStatic
        fun main(args: Array<String>) {
            loadDataFromFile()
            val scanner = Scanner(System.`in`)

            while (true) {
                println("\nOptions:")
                println("1. Add Parking Record")
                println("2. View All Parking Records")
                println("3. Search by License Plate")
                println("4. Search by Owner's Phone")
                println("5. Filter by Parking Hour")
                println("6. Check Payment Status")
                println("7. Exit")
                print("Enter your choice: ")
                val choice = scanner.nextInt()
                scanner.nextLine()

                when (choice) {
                    1 -> addParkingRecord(scanner)
                    2 -> viewAllParkingRecords()
                    3 -> searchByLicensePlate(scanner)
                    4 -> searchByOwnerPhone(scanner)
                    5 -> filterByParkingHour(scanner)
                    6 -> checkPaymentStatus(scanner)
                    7 -> {
                        saveDataToFile()
                        println("Exiting...")
                        System.exit(0)
                    }
                    else -> println("Invalid choice. Please try again.")
                }
            }
        }

        private fun addParkingRecord(scanner: Scanner) {
            println("\nEnter Car Details:")
            print("Brand and Model: ")
            val brandModel = scanner.nextLine()
            print("License Plate: ")
            val licensePlate = scanner.nextLine()
            print("Owner's Phone: ")
            val ownerPhone = scanner.nextLine()
            print("Parking Hour: ")
            val parkingHour = scanner.nextInt()
            print("Prepaid Hours: ")
            val prepaidHours = scanner.nextInt()
            scanner.nextLine()

            val record = CarParkingRecord(brandModel, licensePlate, ownerPhone, parkingHour, prepaidHours)
            records.add(record)
            println("Parking record added successfully.")
        }

        private fun viewAllParkingRecords() {
            if (records.isEmpty()) {
                println("No parking records available.")
            } else {
                println("\nAll Parking Records:")
                records.forEach { println(it) }
            }
        }

        private fun searchByLicensePlate(scanner: Scanner) {
            print("\nEnter License Plate to search: ")
            val licensePlate = scanner.nextLine()
            val foundRecord = records.find { it.licensePlate.equals(licensePlate, ignoreCase = true) }
            if (foundRecord != null) {
                println("Matching Record: $foundRecord")
            } else {
                println("No matching record found.")
            }
        }

        private fun searchByOwnerPhone(scanner: Scanner) {
            print("\nEnter Owner's Phone to search: ")
            val ownerPhone = scanner.nextLine()
            val foundRecord = records.find { it.ownerPhone == ownerPhone }
            if (foundRecord != null) {
                println("Matching Record: $foundRecord")
            } else {
                println("No matching record found.")
            }
        }

        private fun filterByParkingHour(scanner: Scanner) {
            print("\nEnter Parking Hour to filter: ")
            val parkingHour = scanner.nextInt()
            val matchingRecords = records.filter { it.parkingHour == parkingHour }
            if (matchingRecords.isNotEmpty()) {
                println("Matching Records:")
                matchingRecords.forEach { println(it) }
            } else {
                println("No matching record found.")
            }
        }

        private fun checkPaymentStatus(scanner: Scanner) {
            print("\nEnter License Plate to check payment status: ")
            val licensePlate = scanner.nextLine()
            val foundRecord = records.find { it.licensePlate.equals(licensePlate, ignoreCase = true) }
            if (foundRecord != null) {
                val currentHour = getCurrentHour()
                val additionalHours = currentHour - foundRecord.parkingHour
                if (foundRecord.prepaidHours >= additionalHours) {
                    println("No additional payment required. Car can leave.")
                } else {
                    println("Additional payment required for $additionalHours hours.")
                }
            } else {
                println("No matching record found.")
            }
        }

        private fun getCurrentHour(): Int {
            return 10
        }

        private fun saveDataToFile() {
            try {
                val writer = File(FILENAME).printWriter()
                records.forEach { writer.println(it.toText()) }
                writer.close()
                println("Data saved successfully to file.")
            } catch (e: Exception) {
                println("Failed to save data to file: ${e.message}")
            }
        }

        private fun loadDataFromFile() {
            records.clear()
            try {
                val file = File(FILENAME)
                if (file.exists()) {
                    file.forEachLine { line ->
                        val record = CarParkingRecord.fromText(line)
                        record?.let { records.add(it) }
                    }
                    println("Data loaded successfully from file.")
                } else {
                    println("No existing data found. Starting with an empty record list.")
                }
            } catch (e: Exception) {
                println("Failed to load data from file: ${e.message}")
            }
        }
    }
}
